src/compute/src/sink/persist_sink.rs

```diff
@@ -1181,3 +1191,87 @@ where
     let token = Rc::new(shutdown_button.press_on_drop());
     (output_stream, token)
 }
+
+/// This is for REFRESH options on materialized views. It adds an operator that rounds up the
+/// timestamps of data and frontiers to the time of the next refresh. See
+/// `doc/developer/design/20231027_refresh_every_mv.md`.
+///
+/// Note that this currently only works with 1-dim timestamps. (This is not an issue for WMR,
+/// because iteration numbers should disappear by the time the data gets to the Persist sink.)
+fn round_up<G>(
+    coll: Collection<G, Result<Row, DataflowError>, Diff>,
+    refresh_schedule: RefreshSchedule,
+) -> Collection<G, Result<Row, DataflowError>, Diff>
+where
+    G: Scope<Timestamp = Timestamp>,
+{
+    let mut builder = OperatorBuilder::new("round_up".to_string(), coll.scope().clone());
+    let (mut output_buf, output_stream) = builder.new_output();
+    let mut input = builder.new_input_connection(&coll.inner, Pipeline, vec![Antichain::new()]);
+    builder.build(move |capabilities| {
+        let mut capability = Some(capabilities.into_element());
+        let mut buffer = Vec::new();
+        move |frontiers| {
+            let ac = frontiers.into_element();
+            match ac.frontier().to_owned().into_option() {
+                Some(ts) => {
+                    match refresh_schedule.round_up_timestamp(ts) {
+                        Some(rounded_up_ts) => {
+                            capability.as_mut().unwrap().downgrade(&rounded_up_ts);
+                        }
+                        None => {
+                            // We are past the last refresh. Drop the capability to signal that we
+                            // are done.
+                            capability = None;
+                        }
+                    }
+                }
+                None => {
+                    capability = None;
+                }
+            }
+            if let Some(capability) = &capability {
+                input.for_each(|_cap, data| {
```

Comment:
Here it would also be nice to have a comment stating that we can't use `_cap` to get an output session because we might have advanced our output frontier already beyond the frontier of this capability.

Reply:
Good point! I think it's not trivially clear that this operator is correct in all corner cases. Specifically, it relies on properties that `round_up_timestamp` is deterministic, which is not documented as such.

Reply:
Thanks, I've added the comment, and I've also documented that `round_up_timestamp` is monotonic, which I think is the needed property.
```
// Round up the frontier.
// Note that `round_up_timestamp` is monotonic. This is needed to ensure that the timestamp (t) of any
// received data that has a larger timestamp than the original frontier (f) will get rounded up to a time
// that is at least at the rounded up frontier. In other words, monotonicity ensures that
// when `t >= f` then `round_up_timestamp(t) >= round_up_timestamp(f)`.
```